### ISSUE TITLE
Empty ServiceAccount volumes when using Mesos

### ISSUE THREAD

**Issue by MikeSpreitzer** [kubernetes/kubernetes#31062]
BUG REPORT

**Kubernetes version** (use `kubectl version`): release 1.3.5

**Environment**:
- **Cloud provider or hardware configuration**: bare metal
- **OS** (e.g. from /etc/os-release): Ubuntu 14.04
- **Kernel** (e.g. `uname -a`): 3.13.0-92-generic
- **Install tools**: ansible
- **Others**: Mesos 1.0.0-rc2 or 1.0.1-rc1; Docker 1.11.2

Create a cluster with Mesos and Kubernetes as a Mesos framework.

Create a pod with an `ls` command inside a container.  Exec `ls /var/run/secrets/kubernetes.io/serviceaccount/` inside that container.  You will see that the directory is empty.

It looks like the Mesos integration did not pick up the fix to #26943

**Comment by MikeSpreitzer** [comment]
Here is the relevant part of `docker inspect` output for an example pod:
```
        "Mounts": [
            ...,
            {
                "Source": "/var/tmp/mesos/slaves/f7aa3868-ef12-4473-b727-6c5972faaf5a-S0/frameworks/f7aa3868-ef12-4473-b727-6c5972faaf5a-0001/executors/64853556076a8afb_k8sm-executor/runs/e985ac73-e34e-462d-910e-d149ff724d5f/pods/ad945c9c-674b-11e6-90b3-080027ea53f2/volumes/kubernetes.io~secret/default-token-3vbbn",
                "Destination": "/var/run/secrets/kubernetes.io/serviceaccount",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
...
```

And from the host's point of view, `/var/tmp/mesos/slaves/f7aa3868-ef12-4473-b727-6c5972faaf5a-S0/frameworks/f7aa3868-ef12-4473-b727-6c5972faaf5a-0001/executors/64853556076a8afb_k8sm-executor/runs/e985ac73-e34e-462d-910e-d149ff724d5f/pods/ad945c9c-674b-11e6-90b3-080027ea53f2/volumes/kubernetes.io~secret/default-token-3vbbn` is an empty directory.

**Comment by MikeSpreitzer** [comment]
Here is an extract of the executor's log, starting at the first mention of that example container's ID and ending with the first iteration of a repeating pattern of complaints.
[extract.txt](https://github.com/kubernetes/kubernetes/files/428666/extract.txt)

**Comment by MikeSpreitzer** [comment]
One consequence is that the kube DNS application does not work.

**Comment by qiujian16** [comment]
Could you share the flag to start apiserver and controller manager?

**Comment by xingzhou** [comment]
@MikeSpreitzer, I'm looking at this issue, and here is what I did to reproduce the problem:

1. Download the latest Mesos code, build and start 1 single master with 1 single agent.
2. Download the latest Kubernetes code, use sudo KUBERNETES_CONTRIB=mesos make to build.
3. Use the following commands to start apiserver, controller-manager and scheduler respectively:
```
km apiserver   --address=${KUBERNETES_MASTER_IP}   --etcd-servers=http://${KUBERNETES_MASTER_IP}:4001   --service-cluster-ip-range=10.10.10.0/24   --port=8888   --cloud-provider=mesos   --cloud-config=mesos-cloud.conf   --secure-port=0   --admission-control=NamespaceLifecycle,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota,DefaultStorageClass   --service-account-key-file=/home/k8s/k8s/src/k8s.io/kubernetes/cert/server.key   --service-account-lookup=false --bind-address=127.0.0.1   --insecure-bind-address=127.0.0.1   --insecure-port=8888 --v=3 --client-ca-file=/home/k8s/k8s/src/k8s.io/kubernetes/cert/ca.crt --tls-private-key-file=/home/k8s/k8s/src/k8s.io/kubernetes/cert/server.key>apiserver.log 2>&1 &

km controller-manager   --master=${KUBERNETES_MASTER_IP}:8888   --cloud-provider=mesos   --cloud-config=/home/k8s/k8s/src/k8s.io/kubernetes/mesos-cloud.conf    --service-account-private-key-file=/home/k8s/k8s/src/k8s.io/kubernetes/cert/server.key --root-ca-file=/home/k8s/k8s/src/k8s.io/kubernetes/cert/ca.crt   --enable-hostpath-provisioner=false   --v=3 >controller.log 2>&1 &

km scheduler   --address=${KUBERNETES_MASTER_IP}   --mesos-master=${MESOS_MASTER}   --etcd-servers=http://${KUBERNETES_MASTER_IP}:4001   --mesos-user=root   --api-servers=${KUBERNETES_MASTER_IP}:8888   --cluster-dns=10.10.10.10   --cluster-domain=cluster.local   --v=2 >scheduler.log 2>&1 &
```

After all of these steps, I can use 'kubectl create -f pod.yml' to create a single pod and the pod is running. then I checked the "/var/run/secretes/kubernetes.io/serviceaccount" folder and all the token/certificates files are there.  Not sure whether there is anything special that I need to config in my env to reproduce the issue, we can have a discussion on this then. thanks

**Comment by MikeSpreitzer** [comment]
For the record, we are pursuing this in email.

**Comment by MikeSpreitzer** [comment]
For the record, following is the solution I found.  The problem is that when using Mesos, the cluster operator does not launch the kubelet --- so the solutions in #26943 that modify how the kubelet is launched are not applicable.  Instead I modified the way the Mesos slave is launched on each worker node.  I am launching it in a Docker container, and the relevant Docker volume mount is now `--volume=/var/tmp/mesos:/var/tmp/mesos:rw,rshared`.  The addition of that `,rshared` is critical.  But that is not sufficient.  I had to also pre-condition `/var/tmp/mesos` with these two commands (neither alone is sufficient):
```
mount --bind /var/tmp/mesos /var/tmp/mesos
mount --make-rshared /var/tmp/mesos
```
